streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java

```diff
@@ -309,230 +311,275 @@
+     *                     result
+     * @param aggregator   an {@link Aggregator} that computes a new aggregate result
+     * @param named        a {@link Named} config used to name the processor in the topology
+     * @param materialized an instance of {@link Materialized} used to materialize a state store.
+     *                     Cannot be {@code null}.
+     * @param <VR>         the value type of the resulting {@link KTable}
+     * @return a {@link KTable} that contains "update" records with unmodified keys, and values that
+     * represent the latest (rolling) aggregate for each key. If the aggregate function returns
+     * {@code null}, it is then interpreted as deletion for the key, and future messages of the same
+     * key coming from upstream operators will be handled as newly initialized value.
      */
     <VR> KTable<K, VR> aggregate(final Initializer<VR> initializer,
                                  final Aggregator<? super K, ? super V, VR> aggregator,
                                  final Named named,
                                  final Materialized<K, VR, KeyValueStore<Bytes, byte[]>> materialized);
 
     /**
-     * Create a new {@link TimeWindowedKStream} instance that can be used to perform windowed aggregations.
+     * Create a new {@link TimeWindowedKStream} instance that can be used to perform windowed
+     * aggregations.
+     *
      * @param windows the specification of the aggregation {@link Windows}
      * @param <W>     the window type
      * @return an instance of {@link TimeWindowedKStream}
      */
     <W extends Window> TimeWindowedKStream<K, V> windowedBy(final Windows<W> windows);
 
     /**
-     * Create a new {@link SessionWindowedKStream} instance that can be used to perform session windowed aggregations.
+     * Create a new {@link SessionWindowedKStream} instance that can be used to perform session
+     * windowed aggregations.
+     *
      * @param windows the specification of the aggregation {@link SessionWindows}
      * @return an instance of {@link TimeWindowedKStream}
      */
     SessionWindowedKStream<K, V> windowedBy(final SessionWindows windows);
 
+    /**
+     * {@code CogroupedKStream} is an abstraction of multiple <i>grouped</i> record streams of
+     * {@link KeyValue} pairs. It is an intermediate representation of one or more {@link KGroupedStream}s
+     * in order to apply one or more aggregation operations on the original {@link KGroupedStream}
+     * records.
+     * <p>
+     * It is an intermediate representation after a grouping of {@link KStream}s, before the
+     * aggregations are applied to the new partitions resulting in a {@link KTable}.
+     * <p>
+     *
+     * @param aggregator an {@link Aggregator} that computes a new aggregate result
+     */
+    <T> KCogroupedStream<K, V, T> cogroup(final Aggregator<? super K, ? super V, T> aggregator);
```

Comment:
We should update the generic type -- maybe you can to small MINOR PR that fixed generic type naming for all classes (we often use `V`, `V2`, `VR` etc and those are all terrible -- would be worth a small cleanup PR on the side.) (Cf `KStreamFlatTransform` that uses the new/better naming schema already).